server/src/main/java/org/elasticsearch/cluster/NodeConnectionsService.java

```diff
@@ -83,8 +83,7 @@ public void connectToNodes(DiscoveryNodes discoveryNodes) {
         for (final DiscoveryNode node : discoveryNodes) {
             final boolean connected;
             try (Releasable ignored = nodeLocks.acquire(node)) {
-                nodes.putIfAbsent(node, 0);
-                connected = transportService.nodeConnected(node);
+                connected = (nodes.putIfAbsent(node, 0) != null);
```

Comment:
can we avoid going on the management thread if we are truly connected? This will be relevant on the master where where the node is connected to during join validation. Also, I think the name `connected` for the variable is a bit mislead now. Maybe flip it and call it `shouldConnect`? 

It will also good to have a comment explaining the rational for this check.

PS - as discussed, it will be good to explore using this component on non-elected-master nodes which will make my condition tweak obsolete.